semaphore/semaphore.go

```diff
@@ -67,7 +68,7 @@ func (s *Weighted) Acquire(ctx context.Context, n int64) error {
 			// fix up the queue, just pretend we didn't notice the cancelation.
 			err = nil
 		default:
-			s.waiters.Remove(elem)
+			waiterList.Remove(elem)
```

Comment:
elem may moved between `waiterList` and `impossibleWaiters`, so the elem may not be  removed successfully here, cause resource leaking.

Reply:
This is a nice catch actually, here it should remove it from both lists to avoid resource leaking.

Reply:
I updated PR to fix this 👍🏻